### ISSUE TITLE
Implement ability for Flutter tooling to discover isolates in an Android app. Needed by attach command.

### ISSUE THREAD

**Issue by matthew-carroll** [flutter/flutter#22009]
Implement ability for Flutter tooling to discover engine ports for running engines in an Android app. Needed by attach command.

This ability used to exist via BroadcastReceivers but the tooling was out of date and the Android code was removed.

This ability should be reintroduced and the tooling updated.

**Comment by mklim** [comment]
I think this may already be fixed, or maybe I'm misunderstanding the issue. `flutter attach` is already able to automatically discover and work with `examples/flutter_view` for me. 

It looks like `flutter attach` finds a port to attach to by [parsing the device logs](https://github.com/flutter/flutter/blob/master/packages/flutter_tools/lib/src/protocol_discovery.dart#L63) if the port isn't specified at the command line. The add to app flow still logs everything normally, so there isn't an inherent incompatibility there.

**Comment by mklim** [comment]
What do you think @matthew-carroll?

**Comment by matthew-carroll** [comment]
@mklim what we're looking for here is the ability for the tooling to speak directly to the Android app to discover all running engines at a given moment.

It's true that we currently support log parsing but that comes with a problem.  As engines come and go, the logs end up with ports that are no longer available. It's also very tedious for developers to watch logs for this info.

What we'd like to facilitate is the ability for both the IDE and the CLI to essentially say "Hey Android app, list every Flutter engine that's running right now and give me its port."  Then the developer can select which ever engine he/she wants to attach to.

When the CLI is invoked, we can expect that either there is only 1 device available, or the developer has provided the device to target.  From there we should be able to use ADB to send an Intent from the OS that will be received by some part of our app (the part that we need to write).  After that it's worth considering the appropriate way for the responding Activity or BroadcastReceiver to determine which engines are currently running.

**Comment by mklim** [comment]
Got it, thanks for the clarification. So the problem isn't that `flutter attach` doesn't work with Add2App in general. It's that with `Add2App` there could be multiple instances of the engine running at once, and `flutter attach` always just picks one now using log parsing. This issue is to add the option for devs to see all the running instances listed, and then pick one.

**Comment by matthew-carroll** [comment]
Yeah, pretty much.  Initially the goal is to create the capability to discover the active ports.  We can then discuss the best way to represent that at the CLI.  Then the IDE needs to leverage the CLI.

**Comment by mklim** [comment]
This is blocked on flutter/flutter#22750.

**Comment by mklim** [comment]
Looks like the `flutter` tooling is already aware of all the different isolates. See [`FlutterView` in `vmservice.dart`](https://github.com/flutter/flutter/blob/2d3ff10/packages/flutter_tools/lib/src/vmservice.dart#L1401-L1412). That already has a handle to a particular `uiIsolate`. Each connected device has an array of views.

The bad news is that I'm seeing one command broadcast to all available isolates because it looks like generally the `flutter` commands deliberately loop through all devices, and then all FlutterViews, for every command. See [this segment in `run_hot.dart`](https://github.com/flutter/flutter/blob/a02f9d5/packages/flutter_tools/lib/src/run_hot.dart#L433-L443) for a single example of this, but there's instances of that nested loop everywhere in the file.

I think the change we want there is to probably refactor the tooling to take a flag for specific isolates to talk to, and then factor out all these for loops to cycle act on a list that's filtered down to the selected isolate, if present. 

It's also worth noting that in the Observatory it is possible to connect to and debug a specific isolate. But from what I can tell the VSCode debugger runs on all isolates simultaneously. There may be more cases with our tooling like this that handle multiple isolates in different ways that I'm not aware of at the moment.

/cc @DanTup

**Comment by DanTup** [comment]
> It's also worth noting that in the Observatory it is possible to connect to and debug a specific isolate. But from what I can tell the VSCode debugger runs on all isolates simultaneously.

Isolates are mapped to threads in VS Code, so if there are multiple they'll appear in the `Call Stack` section where you can switch between them:

<img width="470" alt="screen shot 2018-10-11 at 8 16 35 am" src="https://user-images.githubusercontent.com/1078012/46786859-091ec000-cd2e-11e8-929a-93348e66f229.png">

Things like the Variables pane, expression evaluation, stepping are based on the selected isolate here.

That said, I'm a little confused here - if multiple Flutter views are running in the same VM, why do we want to attach to one specifically? I think there may be some confusion between "two VMs" and "one VM running two views". If there are multiple VMs, the user needs to be able to choose which one (essentially, which VM service) to connect to; but if there's a single VM running multiple views, I think it should debugged as a single process using the IDEs concept of threads (which I think both IDEs already do).

The issue I was aware of with attach was not about being able to necessarily "select a view" but being able to attach (eg. discover an Observatory port) after the VM has already started (eg. you've missed it's URL in the log stream). If there's always one VM for all flutter views I think the problem is less of "discovering the views" and more about "discovering the VM service we can connect to after we missed its startup"?

**Comment by mklim** [comment]
Thanks for the clarification, thanks. I think what I was missing from that was setting a breakpoint that would just get hit in the one isolate, but I can see a way to work around it now by pausing the extra isolates from that pane.

> If there are multiple VMs, the user needs to be able to choose which one (essentially, which VM service) to connect to; but if there's a single VM running multiple views, I think it should debugged as a single process using the IDEs concept of threads (which I think both IDEs already do).

I think this makes sense. A bit of an important qualification on this though is that it's extremely unlikely to end up in a situation with multiple VMs running at once, from what I've gathered so far. There's only ever one Dart VM per process. Talked about this a bit more in flutter/flutter#22750.

> That said, I'm a little confused here - if multiple Flutter views are running in the same VM, why do we want to attach to one specifically?

The use case this issue is tracking is for if a developer has added Flutter to their existing app and is using separate Flutter isolates to each power different smaller views within their larger UI. They'd want to potentially debug just one of them, reload just one of them, etc, since the multiple different flutter instances aren't necessarily related. Right now the tooling is a little mixed in how it handles this situation, but generally tends to propagate any commands to all the isolates at once.

**Comment by DanTup** [comment]
> They'd want to potentially debug just one of them, reload just one of them, etc, since the multiple different flutter instances aren't necessarily related

Ah, this makes more sense. I was wondering why things like breakpoints would matter if they're running different code, but it seems more important with things like hot reload (and I guess performance overlay toggles, etc.).

That said; if we make the user select an isolate and then filter everything to that isolate (including breakpoints, etc.), what if they've spawned their own isolates and expect to be able to add breakpoints and debug them?

**Comment by matthew-carroll** [comment]
We should avoid thinking of isolates as simply representing "views".  I think we'll see more and more cases moving forward where a given isolate has no UI.  I would recommend that we adjust our thinking to a perspective in which there may be zero or more independent Dart execution contexts.  Some may have UI, some may not.  We'd like the ability for a developer to choose a single isolate and fully debug that one isolate without any noise from other isolates.

**Comment by matthew-carroll** [comment]
@DanTup to the point you just made about child isolates, I guess even the term isolate isn't the best.  Waht we're really talking about is "root isolates".  We'd like to restrict debugging to a "root isolate" which could then include its children.

**Comment by DanTup** [comment]
> We'd like to restrict debugging to a "root isolate" which could then include its children.

Makes sense! Can we tie isolates back to the isolate/root that spawned them? I can't see anything obvious in the VM service docs, but we'd need to know when a new isolate spawned whether we need to consider it part of the group (eg. to send breakpoints for it).

**Comment by DanTup** [comment]
Looking through #23388 I can see how to attach to a specific isolate, but I can't see how to get a list of isolates for a device - is this possible or is there still work to do for this?

**Comment by DanTup** [comment]
> It would be nice if it could automatically detect previously started and running observatory ports, but I don't think that's a directly related issue to the Add2App feature.

I thought that is what this issue was initially intended for based on the original comment (though it did move on to talking about debugging specific views, which I think it what's been solved).

I'll leave @matthew-carroll to clarify - maybe we just need to open another issue for that if there isn't already one. Without it, the workflow is a bit clumsy - if you've launched your Android app and it's displayed a Flutter view and then you decide you want to attach and make changes, I don't think it's possible (I don't think the user can possibly get the Observatory url/port at this point?).

**Comment by mklim** [comment]
> if you've launched your Android app and it's displayed a Flutter view and then you decide you want to attach and make changes, I don't think it's possible (I don't think the user can possibly get the Observatory url/port at this point?).

The port should be in the device logs, same as in a regular app. I may be missing the Add2App specific problem here? You could run `flutter attach` before you first open the app or before the app started any Flutter instances, and then it would attach as soon as the first instance of Flutter ran and logged the Observatory port. It's more of a pain, but if you need to run `flutter attach` after Flutter has already been started on the device you can grep the device logs for the latest message of "Observatory running at..." for the port to use with `flutter attach --debug-port`.

I think it makes sense to file another bug for attach to automatically connect to running ports, it would be easier to use.
